src/communication/global_search_queue_policy.hpp

```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#ifndef NMC_USE_GLOBAL_SEARCH_QUEUE
+#error "global_search_queue_policy.hpp should only be compiled in a NMC_USE_GLOBAL_SEARCH_QUEUE build"
+#endif
+
+#include <util/range.hpp>
+#include "base_communicator.hpp"
+
+namespace nest {
+namespace mc {
+namespace communication {
+
+using nest::mc::util::make_range;
+
+template <typename CommunicationPolicy>
+class global_search_communicator: public base_communicator<CommunicationPolicy> {
+public:
+    using base = base_communicator<CommunicationPolicy>;
+    using typename base::event_queue;
+    using typename base::lt_spike_src;    
+    using base::num_groups_local;
+    using base::base;
+    
+protected:
+    using base::cell_group_index;
+    using base::connections_;
+
+public:
+    // go over each block of connections with the same source, and search for the
+    // associated block of spikes with the same source, and then push the product of events
+    // O(connections/spike * log(spikes))
+    std::vector<event_queue> make_event_queues(const gathered_vector<spike>& global_spikes)
+    {
+        // queues to return
+        auto queues = std::vector<event_queue>(num_groups_local());
+
+        // con_next is the beginning of the current block of connections with
+        // the same source
+        auto con_next = connections_.cbegin();
+        const auto con_end = connections_.cend();
+
+        // spikes_next is the beginning of the current block of spikes with
+        // the same source
+        const auto& spikes = global_spikes.values();
+        auto spikes_next = spikes.cbegin();
+        const auto spikes_end = spikes.cend();
+
+        // Search for next block of spikes and connections with the same sender
+        while (con_next != con_end) {
+            // we grab the next block of connections from the same sender
+            const auto src = con_next->source();
+            const auto targets = std::equal_range(con_next, con_end, src);
```

Comment:
`std::equal_range` does two binary searches, one to get the lower bound and one to get the upper bound.
We know that `con_next` is the lower bound, so in practice you only need the one call to `std::upper_bound` to find the last connection with the same source.